src/sml.cpp

```diff
@@ -288,29 +310,38 @@ void pow(double &val, signed char &scaler)
 
 void smlOBISByUnit(long long int &val, signed char &scaler, sml_units_t unit)
 {
-  unsigned char i = 0, pos = 0, size = 0, y = 0;
+  unsigned char i = 0, pos = 0, size = 0, y = 0, bit = 0;
+  sml_states_t type;
   val = -1; /* unknown or error */
   while (i < listPos) {
     pos++;
-    if (pos == 4 && listBuffer[i + 1] != unit) {
-      /* if unit at position 4 is not 0x1e (wh) return unknown */
+    size = (int)listBuffer[i++];
+    type = (sml_states_t)listBuffer[i++];
+    if (pos == 4 && listBuffer[i] != unit) {
+      /* return unknown (-1) if unit does not match */
       return;
     }
     if (pos == 5) {
-      scaler = listBuffer[i + 1];
+      scaler = listBuffer[i];
     }
     if (pos == 6) {
-      size = (int)listBuffer[i];
       val = 0;
       for (y = 0; y < size; y++) {
         // convert value
         // SML_LOG("size = %i, pos = %i, bit = %i\n", size, y + 1,
         //        (8 * (size - 1)) - (8 * y));
-        val |= (long long int)listBuffer[i + y + 1]
-               << ((8 * (size - 1)) - (8 * y));
+
+        // bitwise NOT for negative numbers
+        bit = (type == SML_DATA_SIGNED_INT && (listBuffer[i] & (1 << 7)))
```

Comment:
I can't comment on much of this PR anymore, but this section seems really strange to me that you start working on bit flips and adjusting the value read. To maybe make the original problem more clear: The Bit Pattern received via SML is a the correct bit pattern for a signed number of whatever bit with the value is in (whch is 16 bit, in my personal example). What your original code did was to paste these bits as the least significant bits into a long long int, which removed the sign bit (if there was one).

What appears to happen here is that you check for the sign bit, then convert the remaining bits into the absolut (posive) value and in line 340/341 convert it back to the negative value.

Not only is this unneccesary, but it now acutally introduces the problem that we need to be aware of whether the value provided via SML is in ones' complement or twos' complement - your code now assumes the value to be ones' complement.

In the past, I have prosposed two approaches: #6 read the value into a correct-sized signed integer value (int16_t in this example) and had the compiler to the conversion - no thinking necessary whatsoever, as long as what the compiler does automatically matches what the meter does. The original #9 then performed sign extension manually, i.e., assumed that the value was some kind of complement representation and adjusted all the bits that were *not* read from the meter - both ones' complement and twos' complement would pass through at least that section of code with an unaltered value.

Reply:
Seems you're right. The bit pattern received via SML is a the correct bit pattern for a signed number. 

Conversion can be done much simpler in code. 

I hope that should do it now ...